pkg/wasm/imagefetcher.go

```diff
@@ -0,0 +1,272 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+	"github.com/hashicorp/go-multierror"
+)
+
+// This file implements the fetcher of "Wasm Image Specificiation" compatible container images.
+// The spec is here https://github.com/solo-io/wasm/blob/master/spec/README.md.
+// Basically, this supports fetching and unpackaging three types of container images containing a Wasm binary.
+
+type ImageFetcherOption struct {
+	Username string
+	Password string
+	// TODO(mathetake) Add signature verification stuff.
+}
+
+func (o *ImageFetcherOption) useDefaultKeyChain() bool {
+	return o.Username == "" || o.Password == ""
+}
+
+type ImageFetcher struct {
+	fetchOpt remote.Option
+}
+
+func NewImageFetcher(opt ImageFetcherOption) *ImageFetcher {
+	var fetchOpt remote.Option
+	// TODO(mathetake): have "Anonymous" option?
+	if opt.useDefaultKeyChain() {
+		// Note that default key chain reads the docker config from DOCKER_CONFIG
+		// so must set the envvar when reaching this branch is expected.
+		fetchOpt = remote.WithAuthFromKeychain(authn.DefaultKeychain)
+	} else {
+		fetchOpt = remote.WithAuth(&authn.Basic{Username: opt.Username})
+	}
+	return &ImageFetcher{
+		fetchOpt: fetchOpt,
+	}
+}
+
+// Fetch is the entrypoint for fetching Wasm binary from Wasm Image Specification compatible images.
+func (o *ImageFetcher) Fetch(url string) ([]byte, error) {
+	ref, err := name.ParseReference(url)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse url in image reference: %v", err)
+	}
+
+	img, err := remote.Image(ref, o.fetchOpt)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch image: %v", err)
+	}
+
+	manifest, err := img.Manifest()
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve manifest: %v", err)
+	}
+
+	if manifest.MediaType == types.DockerManifestSchema2 {
```

Comment:
Can we add some info log when trying to parse each variant of image? So that if there is any error, we can map the error message returned with the image variant.

Reply:
all the returned error contains the information which variant is tried to parse and how it failed (L87 and L105), so maybe it is redundant for emitting log vs emitting error message at callsite (API implementation). 

FWIW, parsing as OCI images is tricky because we cannot know the variant in advance and instead have to do try-catch style parsing so that's the reason why I use the multi error in L105. https://github.com/istio/istio/blob/8d854173484b6e39d3cd9cfdc5df8b8aeaaa05d5/pkg/wasm/imagefetcher_test.go#L244-L246

In any case, I think it would be simpler to allow callers to emit errors or not in one place after calling `Fetch`, given that errors contain per-variant errors(L105) or docker-variant error (L87). 
Also if this doesn't meet the requirement from API impl side, it is easy to revisit this later:) 

sg?

Reply:
Ah yeah, make sense thanks!